lua/autorun/client/cl_devtools.lua

```diff
@@ -1,144 +1,165 @@
+
+        for _, effect in ipairs(allEffects) do
+            if (effect.Folder == targetEffect) then
+                return {
+                    EFFECT = effect
+                }
             end
         end
     end
 }
 
--- Returns a string to print
-local function refreshFile(path, filename)
-    local code = file.Read(path, "MOD")
-
-    -- current filename as class format
-    if SWAMP_DEV.itemStruct then
-        local parentFolder = string.gsub(string.sub(path, 1, -#filename - 2), ".*/", "")
 
-        -- if parent isn't a struct then use parent as class, else use filename
-        if (not SWAMP_DEV.structTypes[parentFolder]) then
-            SWAMP_DEV.structTypes[SWAMP_DEV.itemStruct](parentFolder)
-        else
-            SWAMP_DEV.structTypes[SWAMP_DEV.itemStruct](string.TrimRight(string.gsub(path, "^.*/", ""), ".lua"))
+local rootDirectory = "addons/contrib/"
+local refreshFolder
+local loadFile
+do 
+    local loadFunc = function() end
+    function loadFile(filePath, environmentVars)
+        local code = file.Read(filePath, "MOD")
+        if not code then error("No file exists with the file path: " .. filePath) end
+        environmentVars = environmentVars or {}
+        loadFunc = CompileString(code, "SwapDevTools/" .. filePath)
+        -- setup our environment
+        local environment = _G
+        for key,var in pairs(environmentVars) do
+            environment[key] = var
         end
-    end
+        
+        debug.setfenv(loadFunc,environment)
+
+        xpcall(loadFunc, function(err)
+            print("[contrib] Error during loading file with devtools:", err)
+            debug.Trace()
+        end)
 
-    BaseGamemode = engine.ActiveGamemode()
-    GM = baseclass.Get(BaseGamemode)
-
-    -- handle include
-    for s in string.gmatch(code, "include%s*%b()") do
-        local reqInclude = string.sub(string.gsub(s, "[%s%(%)\'\"]+", ""), 8)
-        local includePath = string.sub(path, 1, -#filename - 1) .. reqInclude
-
-        -- First check the parent folder
-        if (file.Exists(includePath, "MOD")) then
-            print("INCLUDING: " .. refreshFile(includePath, string.gsub(reqInclude, ".*/", "")))
-            -- Then check the lua/ game folder
-        elseif (file.Exists("addons/contrib/lua/" .. reqInclude, "MOD")) then
-            print("INCLUDING: " .. refreshFile("addons/contrib/lua/" .. reqInclude, string.gsub(reqInclude, ".*/", "")))
-        else
-            print("Include handled by file")
+        -- remove the variaables from the global.
+        for key,_ in pairs(environmentVars) do
+            environment[key] = nil
         end
     end
 
-    -- only replace includes that don't use variables
-    RunString(string.gsub(code, "include%s*%(%s*[\'\"].-[\'\"]%s*%)", ""), "swampDevTools")
-
-    return filename
-end
-
-local function recurseRefresh(root, times, curPath)
-    local files, folders = file.Find(root .. curPath .. "*", "MOD")
+    local fileTime = file.Time
```

Comment:
What's the point of this?

Reply:
Localization reduces indexing times, as this is called frequently this will be slightly more efficient.

Reply:
I really don't think the performance impact will matter. Code is cleaner without it.

Reply:
You can leave it in if you want, but I think the overwhelming majority of the lag from this function will be due to the file.Time call looking at the disk synchronously.

If you're really worried about performance, what you could do is keep track of the 5 most recently edited files and update them 3x per second, and then update other files every 5 seconds, and additionally, don't do them all at once, spread them out over the 5 seconds and do a few every frame.